### PR TITLE
Rework entrypoints

### PR DESCRIPTION
**Pull request by qwerty287**
Closes https://github.com/woodpecker-ci/woodpecker/discussions/3263
Closes https://github.com/woodpecker-ci/woodpecker/issues/3477

### CODE REVIEW

**Comment by zc-devs** [comment]
LGTM. Build image, please, I'll test in Kubernetes.

File: docs/docs/20-usage/20-workflow-syntax.md
Comment (zc-devs) [reply docs/docs/20-usage/20-workflow-syntax.md:165]: > How are other CIs (gitlab / drone) solving this?

GitHub
1. Uses [just script](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-running-an-inline-python-script) without commands.
2. You can [override shell](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell)
3. For containers you can set [entrypoint](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepswithentrypoint) and [args](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepswithargs).
4. You can use actions (plugins in WP terminology)

GitLab
1. Supports only [scriptinng](https://docs.gitlab.com/ee/ci/yaml/#script). [The image where your job runs must have a working shell in its operating system PATH](https://docs.gitlab.com/runner/executors/docker.html#configure-images-and-services).
2. You can [choose shell](https://docs.gitlab.com/runner/shells), but `the shells don’t have any configuration options`.
3. [How the runner creates Kubernetes pods](https://docs.gitlab.com/runner/executors/kubernetes.html#how-the-runner-creates-kubernetes-pods).

Comment (zc-devs) [reply docs/docs/20-usage/20-workflow-syntax.md:165]: > makes the usage especially for newcomers harder
> adjust the script part somehow / change the way commands are treated as soon as we set an entrypoint

There is enough magic already. Especially for newcomer (say as ex-newcomer).
Apart from being stuck with `kaniko` image, there was another funny moment:
```shell
> docker run -it ubuntu
Status: Downloaded newer image for ubuntu:latest
root@80812effb84a:/# echo $SHELL
/bin/bash
root@80812effb84a:/# test=(1 2 4 8 16 32 64 128)
root@80812effb84a:/# echo ${test[@]}
1 2 4 8 16 32 64 128
```
[Pipeline](https://woodpecker.test.smthd.com/repos/1/pipeline/131)
```yaml
skip_clone: true
steps:
  bash:
    image: ubuntu
    commands:
      - echo $SHELL
      - test=(1 2 4 8 16 32 64 128); echo $${test[@]};
```
```
+ echo $SHELL
/bin/sh
+ test=(1 2 4 8 16 32 64 128); echo ${test[@]};
/bin/sh: 18: Syntax error: "(" unexpected
```

`echo $SHELL` is hint from experienced user 😆

Comment (anbraten) [reply docs/docs/20-usage/20-workflow-syntax.md:165]: Thanks for the github & gitlab links. Super interesting to read through. Had a look at gitlabs code already, but was having a quite tough time finding all the parts. 

This pre-processor using bash syntax is so weird. I'm always running into that trap. Like why is `$SHELL` even working in that example 🤦🏾 

```
+ echo $SHELL
/bin/sh
```

Comment (zc-devs) [reply docs/docs/20-usage/20-workflow-syntax.md:165]: Pre-processor is another story. Usually I use scripts in files.
I was confused by why in Docker Bash syntax (arrays, etc.) works, but in WP doesn't. It is the same image! But there is nuance :) Reading through the code I understood everything.

**Review by 6543 — CHANGES_REQUESTED** [review]
I'm also against adding another config flat

File: docs/docs/20-usage/20-workflow-syntax.md
Comment (6543) [reply docs/docs/20-usage/20-workflow-syntax.md:165]: I'm also against adding another config flat ...
... I want to get rid of the script generation witch prevents it atm and send the commands directly into the container - so this do requery backend changes (improvements)

**Comment by 6543** [comment]
what about, if the custom entrypint is set, thread comments as arguments.

this way you can make both usages work:

```yaml
entrypoint:
  - /usr/bin/curl
 commands:
  - "-v"
  - "--insecure"
  - https://google.com
```

```yaml
entrypoint:
  - /bin/bash
 commands:
  - "-x"
  - "-e"
  - "-c 'echo Hello'; printenv"
```

https://github.com/woodpecker-ci/woodpecker/discussions/3263#discussioncomment-8233224

**Comment by zc-devs** [comment]
> if the custom entrypint is set, thread comments as arguments

That's an option. What's about "magic" aspect then?

Also, we have to find and provide entrypoint for each "scratch" image. We cannot use default image's entrypoint:
```yaml
steps:
  build:
    image: gcr.io/kaniko-project/executor:v1.18.0
    args:
      - --context
      - ./
      - --destination
      - harbor.domain.net/project/image:123
```

Anyway, this would be better than ~nothing~ current state.

**Comment by 6543** [comment]
:point_up: should my soulution not work in this way also ?!?

**Comment by anbraten** [comment]
> there shouldn't be a script anymore. In this case, this could be implemented in a similar way.

We / I had a look at that approach some time ago as well, but actually stopped as there seems to be no smart way of doing this. My first idea was to simply execute commands one by one using sth like `docker exec ...`, but that doesn't work as vars or the current directory wouldn't work anymore. The other idea I had was using `screen` / `tmux`, but that would require even more specific images to be used.

File: docs/docs/20-usage/20-workflow-syntax.md
Comment (zc-devs) [reply docs/docs/20-usage/20-workflow-syntax.md:165]: > How are others solving this?

_Tekton_ can runs [scripts](https://tekton.dev/docs/getting-started/tasks/) as well as [commands/args](https://tekton.dev/docs/pipelines/tasks/#configuring-a-task).
_Argo Workflows_ runs [commands/args](https://github.com/argoproj/argo-workflows/blob/main/examples/hello-hybrid.yaml).

**Comment by zc-devs** [comment]
> Would it be possible to limit entrypoint to the first element of the entrypoint YAML slice?

Of course, yes.

1. We have to find and provide entrypoint for each "scratch" image. We cannot use default image's entrypoint. (https://github.com/woodpecker-ci/woodpecker/pull/3269#issuecomment-1911091750, https://github.com/woodpecker-ci/woodpecker/discussions/3263#discussioncomment-8233224)
2. Would you allow to use commands with entrypoint? Could you override shell (https://github.com/woodpecker-ci/woodpecker/discussions/3263#discussioncomment-8233224, propose, point 1), but using commands as before(script)?
```
entrypoint:
  - /bin/python
  - "-strict"
 commands:
  - print("Hello")
  - quit()
```

**Comment by zc-devs** [comment]
Seems, doesn't work.

Pipeline:
```yaml
skip_clone: true
steps:
  server:
    image: quay.io/curl/curl
    entrypoint:
      - /usr/bin/curl
      - -v
      - google.com
```
Output:
```
curl: try 'curl --help' or 'curl --manual' for more information
```
Warnings:
```
[linter]woodpecker: stepsMust validate one and only one schema (oneOf)
[linter]woodpecker: steps.serverAdditional property entrypoint is not allowed
[bad_habit]Please set an event filter on all when branches
```
[pod-entrypoint-only.yaml](https://github.com/woodpecker-ci/woodpecker/files/14735994/pod-entrypoint-only.yaml.txt)

**Comment by zc-devs** [comment]
[Fixed_custom_entrypoint.patch](https://github.com/woodpecker-ci/woodpecker/files/14737030/Fixed_custom_entrypoint.patch)

1.
```yaml
skip_clone: true
steps:
  server:
    image: quay.io/curl/curl
    entrypoint:
      - /usr/bin/curl
      - -v
      - -k
      - https://google.com
```
```
 0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host google.com:443 was resolved.
* IPv6: 2a00:1450:4007:80e::200e
...
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://google.com/
...
* Connection #0 to host google.com left intact
<HTML><HEAD><meta http-equiv="content-type" content="text/html;charset=utf-8">
<TITLE>301 Moved</TITLE></HEAD><BODY>
<H1>301 Moved</H1>
The document has moved
<A HREF="https://www.google.com/">here</A>.
```
[pod-entrypoint-only.yaml](https://github.com/woodpecker-ci/woodpecker/files/14737035/pod-entrypoint-only.yaml.txt)
[woodpecker-agent-entrypoint-only.log](https://github.com/woodpecker-ci/woodpecker/files/14737039/woodpecker-agent-entrypoint-only.log)

2.
```yaml
skip_clone: true
steps:
  server:
    image: debian
    entrypoint:
      - /bin/bash
      - -c
      - echo $CI_SCRIPT | base64 -d | /bin/bash -e
    commands:
      - echo $SHELL
      - test=(1 2 4 8 16 32 64 128); echo $${test[@]};
```
```
+ echo $SHELL
/bin/sh
+ test=(1 2 4 8 16 32 64 128); echo ${test[@]};
1 2 4 8 16 32 64 128
```
[pod-custom-shell.yaml](https://github.com/woodpecker-ci/woodpecker/files/14737049/pod-custom-shell.yaml.txt)
[woodpecker-agent-custom-shell.log](https://github.com/woodpecker-ci/woodpecker/files/14737050/woodpecker-agent-custom-shell.log)

3.
```yaml
skip_clone: true
steps:
  server:
    image: debian
    commands:
      - echo $SHELL
      - test=(1 2 4 8 16 32 64 128); echo $${test[@]};
```
```
+ echo $SHELL
/bin/sh
+ test=(1 2 4 8 16 32 64 128); echo ${test[@]};
/bin/sh: 18: Syntax error: "(" unexpected
```
[pod-sh.yaml](https://github.com/woodpecker-ci/woodpecker/files/14737054/pod-sh.yaml.txt)
[woodpecker-agent-sh.log](https://github.com/woodpecker-ci/woodpecker/files/14737055/woodpecker-agent-sh.log)

**Comment by anbraten** [comment]
So to summarize the current changes: You can either use commands or entrypoints. If entrypoints is used the part with `/bin/bash - -c  - echo $CI_SCRIPT | base64 -d | /bin/bash -e` is simply skipped, correct?

**Comment by anbraten** [comment]
>  you requested changes, but your concerns were addressed

I could dismiss them if you think that he would be fine with the new changes?

**Comment by anbraten** [comment]
> So the reason for requesting changes (https://github.com/woodpecker-ci/woodpecker/pull/3269#pullrequestreview-1844754093) was the additional config option, that one is resolved, however, the solution in https://github.com/woodpecker-ci/woodpecker/pull/3269#issuecomment-1911073603 is not how it's done now as entrypoint and arguments are in one config option now… So I can't tell you

In that case it would be fair to wait for the approval.

**Review by 6543 — APPROVED** [review]
this will make it so we can not pass arguments ... but It's probably the best we can do as of now :/

**Pull request merged**
